File: Python/Day01/day01.py
def calcular_distancia_total(lista_izquierda, lista_derecha):
    # Ordenar ambas listas
    lista_izquierda_ordenada = sorted(lista_izquierda)
    lista_derecha_ordenada = sorted(lista_derecha)
    
    # Calcular la suma de las diferencias absolutas
    distancia_total = sum(abs(a - b) for a, b in zip(lista_izquierda_ordenada, lista_derecha_ordenada))
    
    return distancia_total



def leer_fichero(archivo):
    """
    Lee un archivo con pares de números separados por espacios y devuelve dos listas
    :param nombre_archivo: Nombre del archivo que contiene los números
    :return: Una tupla de dos listas (numeros_izquierda, numeros_derecha)
    """
    # Listas para almacenar los números de la izquierda y de la derecha
    numeros_izquierda = []
    numeros_derecha = []

    # Leer el archivo línea por línea
    with open(archivo, 'r') as f:
        for linea in f:
            # Separar los números en cada línea
            if linea.strip():  # Ignorar líneas vacías
                num_izquierda, num_derecha = map(int, linea.split())
               
                numeros_izquierda.append(num_izquierda)
                numeros_derecha.append(num_derecha)

    return numeros_izquierda, numeros_derecha


# # Ejemplo de uso
# lista_izquierda = [3, 4, 2, 1, 3, 3]
# lista_derecha = [4, 3, 5, 3, 9, 3]

archivo = "input.txt"

lista_izquierda, lista_derecha = leer_fichero(archivo)

distancia = calcular_distancia_total(lista_izquierda, lista_derecha)

print(f"La distancia total mínima es: {distancia}")

File: Node/Day01/day01.js
const fs = require('fs');
const archivo = 'input.txt';
const DBG = true;

const leerArchivo = (archivo) => {
    return new Promise((resolve, reject) => {
        // Leer las listas desde el archivo
        fs.readFile(archivo, 'utf8', (err, data) => {
            if (err) {
                console.error('Error al leer el archivo:', err);
                return reject(err);
            }

            // Procesar las líneas del archivo
            const lineas = data.split('\n').map(linea => linea.trim());
            const listaDerecha = [];
            const listaIzquierda = [];

            //DBG && console.log('Líneas:', lineas);

            for (const linea of lineas) {
                if (!linea) {
                    console.error('El archivo contiene una línea vacía.');
                    continue;
                }

                [izq, der] = linea.split('   ').map(Number);

              //  DBG && console.log('izq, der:', izq, der);

                listaIzquierda.push(izq);
                listaDerecha.push(der);

                // DBG && console.log('Lista izquierda:', listaIzquierda);
                // DBG && console.log('Lista derecha:', listaDerecha);
            }
            resolve({ listaIzquierda, listaDerecha });

        });
    });
}


const calcularDistancia = (listaIzquierda, listaDerecha) => {
    let distancia = 0;
    const sortedIzquierda = listaIzquierda.sort((a, b) => a - b);
    const sortedDerecha = listaDerecha.sort((a, b) => a - b);

    for (let i = 0; i < sortedIzquierda.length; i++) {
        distancia += Math.abs(sortedIzquierda[i] - sortedDerecha[i]);
    }
    return distancia;
}

const calcularPuntuacionSimilitud=(listaIzquierda, listaDerecha) => {
    // Crear un objeto para contar las ocurrencias en la lista derecha
    const conteoDerecha = listaDerecha.reduce((contador, numero) => {
        contador[numero] = (contador[numero] || 0) + 1;
        return contador;
    }, {});

    // Calcular la puntuación de similitud
    let puntuacionTotal = 0;
    for (const numero of listaIzquierda) {
        puntuacionTotal += numero * (conteoDerecha[numero] || 0);
    }

    return puntuacionTotal;
}


(async () => {
    const { listaIzquierda, listaDerecha } = await leerArchivo(archivo);

    // DBG && console.log('Lista izquierda:', listaIzquierda);
    // DBG && console.log('Lista derecha:', listaDerecha);

    const distancia = calcularDistancia(listaIzquierda, listaDerecha);
    console.log('Distancia:', distancia);

    const puntuacion = calcularPuntuacionSimilitud(listaIzquierda, listaDerecha);
    console.log('Puntuación de similitud:', puntuacion);
})();

